pkg/virt-operator/kubevirt.go

```diff
@@ -0,0 +1,527 @@
+			if err != nil {
+				log.Log.Errorf("Failed to set condition: %v", err)
+			}
+			return err
+		}
+
+		if c.stores.AllEmpty() {
+			// deletion successful
+			err = util.UpdatePhase(kv, v1.KubeVirtPhaseDeleted, c.clientset)
+			if err != nil {
+				log.Log.Errorf("Failed to update phase: %v", err)
+			}
+			err = util.RemoveConditions(kv, c.clientset)
+			if err != nil {
+				log.Log.Errorf("Failed to update condition: %v", err)
+			}
+			err = util.RemoveFinalizer(kv, c.clientset)
+			if err != nil {
+				log.Log.Errorf("Failed to remove finalizer: %v", err)
+			}
+
+			log.Log.Info("KubeVirt deleted")
+			return nil
+		}
+
+		log.Log.Info("Processed deletion for this round")
+		return nil
+	}
+
+	logger.Infof("Handling deployment")
+
+	if kv.Status.Phase == v1.KubeVirtPhaseDeployed {
+		log.Log.Info("Is already deployed")
+		return nil
+	}
+
+	// Set versions...
+	if kv.Status.OperatorVersion == "" {
+		err = util.SetVersions(kv, c.config, c.clientset)
+		if err != nil {
+			log.Log.Errorf("Failed to set versions: %v", err)
+			return err
+		}
+	}
+
+	// Set phase to deploying
+	err = util.UpdatePhase(kv, v1.KubeVirtPhaseDeploying, c.clientset)
+	if err != nil {
+		log.Log.Errorf("Failed to update phase: %v", err)
+		return err
+	}
+
+	// check if there is already an active KubeVirt deployment
+	// TODO move this into a new validating webhook
+	kvs := c.kubeVirtInformer.GetStore().List()
+	for _, obj := range kvs {
+		if fromStore, ok := obj.(v1.KubeVirt); ok {
+			if fromStore.UID == kv.UID {
+				continue
+			}
+			if isKubeVirtActive(&fromStore) {
+				logger.Warningf("There is already a KubeVirt deployment!")
+				err := util.UpdateCondition(kv, v1.KubeVirtConditionSynchronized, k8sv1.ConditionFalse, ConditionReasonDeploymentFailedExisting, "There is an active KubeVirt deployment", c.clientset)
+				if err != nil {
+					log.Log.Errorf("Failed to set condition: %v", err)
+				}
+				return nil
+			}
+		}
+	}
+
+	// add finalizer to prevent deletion of CR before KubeVirt was undeployed
+	err = util.AddFinalizer(kv, c.clientset)
+	if err != nil {
+		log.Log.Errorf("Failed to add finalizer: %v", err)
+		util.UpdateCondition(kv, v1.KubeVirtConditionSynchronized, k8sv1.ConditionFalse, ConditionReasonDeploymentFailedError, fmt.Sprintf("Failed to add finalizer: %s", err), c.clientset)
+		return err
+	}
+
+	// deploy
+	objectsAdded, err := creation.Create(kv, c.config, c.stores, c.clientset, &c.kubeVirtExpectations)
+
+	if err != nil {
+		// deployment failed
+		err := util.UpdateCondition(kv, v1.KubeVirtConditionSynchronized, k8sv1.ConditionFalse, ConditionReasonDeploymentFailedError, fmt.Sprintf("An error occurred during deployment: %v", err), c.clientset)
+		if err != nil {
+			log.Log.Errorf("Failed to set condition: %v", err)
+		}
+		return err
+	}
+
+	if objectsAdded == 0 {
+		// deployment successful
+		err = util.UpdatePhase(kv, v1.KubeVirtPhaseDeployed, c.clientset)
```

Comment:
What I pointed out about the delete related PATCHEs applies here on update as well.  If updating the phase passes, but the condition update fails, then the conditions never get removed. 

Status updating needs to be a single operation for each iteration of the execution loop. Take a look at the updateStatus function in vmi.go.  We're taking care of conditions and phase updates in one operation.